### DISCUSSION TITLE
Add Copilot Plugin

### DISCUSSION THREAD

**Discussion by 7flash** [helix-editor/helix#11824] [Ideas]

**Comment by nick887** [comment]
is there any discussions? i think it is a really useful feature if we can use github copilot in helix and i will choose switch from goland to helix to get my work done

**Comment by sudormrfbin** [comment]
There are no plans to have copilot in the editor core, so this will have to wait till there is proper plugin support.

**Comment by roehst** [comment]
Waiting on this feature to move from neovim to helix for good

**Comment by lukepighetti** [comment]
Has anyone tried integrating this copilot LSP? https://github.com/TerminalFi/LSP-copilot/blob/master/language-server/package.json#L4

**Comment by yudjinn** [comment]
isnt github getting sued for how copilot takes data? I def dont think copilot should be anywhere near core

**Comment by gaetschwartz** [comment]
Is there any update on this ?

**Comment by lukepighetti** [comment]
This seems like something we can solve with LSP integration instead of the plugin system. I believe that's how folks are using it in Sublime https://forum.sublimetext.com/t/github-copilot-for-sublime-text-4-is-coming/64449/3

**Reply by devmanuelli** [reply]
@lukepighetti checkout my pr https://github.com/helix-editor/helix/pull/14876

**Comment by kanwarujjaval** [comment]
Was anyone able to use the [copilot LSP](https://github.com/TerminalFi/LSP-copilot/) successfully?
It would be really useful specially when #2507 is merged.

**Comment by leeola** [comment]
With Copilot X's recent announcement i had to look this up for Helix, landing me here. The LSP idea sounds alright, but it would be nice to craft a deeper UX around Copilot.

Are plugin-like APIs available in Helix to, perhaps, temporarily fork Helix and integrate a Copilot Plugin directly into the forked Helix binary? Ie write a plugin directly into fork of Helix proper, even though we only intend to migrate it to a Plugin asap?

If Copilot X is useful _(big if, heh)_ it would be nice to have a good experience in Helix for that. They have a [NeoVim plugin](https://github.com/github/copilot.vim), for comparison.

**Comment by patrick-kidger** [comment]
Also ended up here after the Copilot X announcement, FWIW. No idea if implementing it as just an LSP is technically feasible, but that would certainly be nice if so.

**Comment by mikkelfj** [comment]
The way things are going with AI these days, some level of non-trivial support would eventually be necessary, but some aspects like documentation could also belong to the build system.

Either way, I suspect we will see competitors, notably also fully open source, to Copilot-X in the coming years (or days, maybe). Also the feature set will evolve drastically.

For these reasons I think both some sense of urgency and restraint is necessary at the same time.

**Comment by 7flash** [comment]
A temporary solution I have for myself is a chatgpt window running a "bridge" script in devtools. It allows chatgpt to communicate  with other apps through a locally running database/job queue. If it makes sense, I can try to make it into a plugin.

**Comment by ptman** [comment]
Not just copilot, there are now several similar services. Copilot, tab9, codeium, codegeex, (vim-ai just uses chatgpt), ...

**Comment by 7flash** [comment]
> A temporary solution I have for myself is a chatgpt window running a "bridge" script in devtools. It allows chatgpt to communicate with other apps through a locally running database/job queue. If it makes sense, I can try to make it into a plugin.

Relevant: https://github.com/kazuki-sf/YouTube_Summary_with_ChatGPT/issues/9

**Comment by ptman** [comment]
And also amazon code whisperer. I suggest title is rewritten to reflect the variety of alternatives

**Comment by Neugierdsnase** [comment]
What can we do to make this happen as soon as possible? I want to switch over completely, but the reality is, that I just don't want miss AI tools.

I'm willing to invest time in this.

**Comment by 7flash** [comment]
> What can we do to make this happen as soon as possible? I want to switch over completely, but the reality is, that I just don't want miss AI tools.
> 
> I'm willing to invest time in this.

Consider running a browser extension which establishes communication between chatgpt and helix, here is minimal example I made in between two instances of chatgpt, but you can imagine to replace second instance with helix editor: https://github.com/7flash/AutoChatGPT

**Comment by rnarenpujari** [comment]
Related discussion thread: https://github.com/helix-editor/helix/discussions/4037

**Comment by kirawi** [comment]
I noticed that the related fork was not linked: https://github.com/helix-editor/helix/pull/6865
AFAIK it works and some people are using it, but we have no plans to merge it into Helix for reasons discussed there

**Comment by Neugierdsnase** [comment]
> I noticed that the related fork was not linked: #6865 AFAIK it works and some people are using it, but we have no plans to merge it into Helix for reasons discussed there

Thanks for linking!

**Comment by 0x61nas** [comment]
> isnt github getting sued for how copilot takes data? I def dont think copilot should be anywhere near core

\+ not everyone wanna be distracted by its stupid suggestions. We should wait for the plugin system to be ready, and am pretty sure someone will write a plugin for this.

**Comment by hemedani** [comment]
how about [codeium](https://codeium.com/)?

**Comment by leona** [comment]
In-case anybody is still looking for this https://github.com/helix-editor/helix/discussions/4037#discussioncomment-8227690

**Comment by tirithen** [comment]
As suggested here https://github.com/helix-editor/helix/discussions/9369 I believe that adding Ollama LSP support would be a great possibility. Although requiring a GPU with enough memory for he heavier models they all run locally on your machine. This might be considered a much safer setup.

Having gpt text completion would be a very welcome addition to helix, and having it work with open locally run models even better.

**Comment by iocron** [comment]
> As suggested here #9369 I believe that adding Ollama LSP support would be a great possibility. Although requiring a GPU with enough memory for he heavier models they all run locally on your machine. This might be considered a much safer setup.
> 
> Having gpt text completion would be a very welcome addition to helix, and having it work with open locally run models even better.

I agree except the heavier models. We don't necessarily need them, because there are a couple great and efficient ollama code models out there without using a big gpu :) https://ollama.ai/library?q=code

**Comment by mikkelfj** [comment]
ollama run mistral
works fine on mac studio M2 Max 32GB RAM, its a 7B model.

**Comment by mikkelfj** [comment]
However, it only really gets interesting when the models can understand content referenced in compile_commands.json

**Comment by tirithen** [comment]
Regardless the model size, the great thing with ollama is that projects like that makes it possible to easily pull the model of your choice, that fits your need and system capabilities while ensuring privacy.

I wonder if there is already good LSPs out there for ollama that could be hooked into helix? Has anyone considered the llm-ls project? Just found it now, but it is supposed to work as an LSP and it can bridge over to LLM runners ollama being one of them.

There has already been some interest in figuring out the LSP integration from @hemedani and @webdev23 on the issue https://github.com/huggingface/llm-ls/issues/49 . Maybe there is someone from here that can help them out with some directions to get going?

**Comment by leona** [comment]
> Regardless the model size, the great thing with ollama is that projects like that makes it possible to easily pull the model of your choice, that fits your need and system capabilities while ensuring privacy.
> 
> I wonder if there is already good LSPs out there for ollama that could be hooked into helix? Has anyone considered the llm-ls project? Just found it now, but it is supposed to work as an LSP and it can bridge over to LLM runners ollama being one of them.
> 
> There has already been some interest in figuring out the LSP integration from @hemedani and @webdev23 on the issue [huggingface/llm-ls#49](https://github.com/huggingface/llm-ls/issues/49) . Maybe there is someone from here that can help them out with some directions to get going?

It's possible to overwrite the openai endpoint of my language server [here](https://github.com/leona/helix-gpt) but if it doesn't follow the openai pattern it won't work, and I don't think ollama does. If people wanted support for ollama it would be fairly easy to add, but any time I've tried locally hosted models they've been pretty poor for this use case.

**Comment by tirithen** [comment]
@leona  thanks for sharing! Your project could be a nice template for running against ollama. When I tested Mistral 7B separately for code assistance in Rust it worked pretty well for me, at least better than without, but I would probably want to run the larger models for even better responses.

I'm personally more into these open models mainly for privacy reasons, but anyhow your project could also be useful as is for the ones that that started the issue and where interested in having copilot running (sorry for hijacking the thread for ollama things by the way).

Really nice with a running LSP plugin example, now I which I just had more time to try this out with ollama. If anyone gets started I'll definitely try to find some time to help out (but preferably in Rust in that case).

**Comment by mikkelfj** [comment]
I'm not quite ready for relying on LLMs for coding: This is mistral 7B

"An octahedron consists of eight vertices and eight triangular faces, with each face being made up of three vertices. Therefore, there are indeed a total of 8 x 3 = 24 individual vertices, but since a triangle is defined by three non-unique vertices, there are only 12 unique triangles in an octahedron."

**Comment by 7flash** [comment]
Hi! I would like to share here my current workflow I found myself comfortable with.

https://github.com/7flash/helix-chatgpt

Notice, it works especially well with Warp Terminal, where I created a workflow/button to execute the script, but as well you can define bash function etc.

What happens when you run the script, it opens a new file with helix where you can write your prompt, and I found it being more comofrtable than any existing UI.

* You can separate user and assistant messages with |user| and |assistant| separators directly in your prompt
* You can reference local files with file:/ references and web links with http[s]:// and their content will be embedded into the prompt
* Now, you obviously won't have to embed the whole content of given webpage into your prompt and therefore, you can specify CSS selector after # hashtag in its path.
* As well, you won't have to embed full source files, so you can reference specific sections using special comments starting with "."

Just sharing here what works for me, but there isn't good documentation yet, so please feel free to contribute.

**Comment by kyfanc** [comment]
Hi everyone! 

@leona thanks for your effort in creating helix-gpt. I made an [attempt](https://github.com/leona/helix-gpt/pull/42) to integrate with ollama. @tirithen may be you would be interested in having a try with it? 

Regarding [llm-ls](https://github.com/huggingface/llm-ls), I believe they are implementing the features as LSP custom methods, which helix built-in LSP client do not support currently.

**Reply by KiaraGrouwstra** [reply]
i tried looking this up for a bit, [looks like](https://github.com/huggingface/llm-ls/blob/59febfea525d7930bf77e1bae85b411631d503e4/crates/llm-ls/src/main.rs#L775-L777):

- `llm-ls/getCompletions`
- `llm-ls/acceptCompletion`
- `llm-ls/rejectCompletion`

would you perhaps know what it would take to allow such custom LSP methods in helix?

**Comment by tirithen** [comment]
@kyfanc I'm off for a trip for a while now without my GPU, but I'll give it a try once I'm back.

Ideally in the end I think this sort of application should be a simple binary rather than a bun + typescript, to be fast and efficient, written in Rust or similar. But I also saw that it at least can build to a binary via bun (I suppose it means bundling and running a full V8 to run the lsp).

Nice that there is some progress either way! :-) And also nice with the approach to support both cloud gpt providers and ollama so the users can choose what fits them best. :-)

**Comment by RayyanNafees** [comment]
>  (I suppose it means bundling and running a full V8 to run the lsp).


@tirithen bun doesnt use v8 actually, it's JavaScript-core which has smaller size

**Comment by llamaha** [comment]
So I really wanted to see this feature (albeit for Claude) but I'm actually seeing good results using aichat: https://github.com/sigoden/aichat

Using this with the "insert-output" and "append-output" commands in Helix the output will write into Helix:

```
:insert-output aichat -r %code% Write a Rust function that will parse an i32 integer and return a result
```

Can run it outside of Helix just as easily:

```
aichat -f src/main.rs what does this code do?
```

Using the shell_pipe feature you can also run aichat against highlighted text

```
:pipe aichat -r %code% please write unit tests for this highlighted function 
```

Or this is also very easy using the command line by specifying the function:

```
aichat -f src/main.rs -r %code% please write unit tests for the function run
```

Note that I've been verbose on purpose for demonstration purposes.  The above commands can be reduced in length and there are hotkeys for the shell commands.

**Comment by mikkelfj** [comment]
The benefit is when AIs get larger contexts and can understand compile_commands.json or compile_flags.txt

And then you can just say, add test cases for the functions below, and add a bash script to drive it. Also write a bat file for windows like the other scripts.

https://clang.llvm.org/docs/JSONCompilationDatabase.html

**Comment by rodrigopim** [comment]
> So I really wanted to see this feature (albeit for Claude) but I'm actually seeing good results using aichat: https://github.com/sigoden/aichat
> 
> Using this with the "insert-output" and "append-output" commands in Helix the output will write into Helix:
> 
> ```
> :insert-output aichat -r %code% Write a Rust function that will parse an i32 integer and return a result
> ```
> 
> Can run it outside of Helix just as easily:
> 
> ```
> aichat -f src/main.rs what does this code do?
> ```
> 
> Using the shell_pipe feature you can also run aichat against highlighted text
> 
> ```
> :pipe aichat -r %code% please write unit tests for this highlighted function 
> ```
> 
> Or this is also very easy using the command line by specifying the function:
> 
> ```
> aichat -f src/main.rs -r %code% please write unit tests for the function run
> ```
> 
> Note that I've been verbose on purpose for demonstration purposes. The above commands can be reduced in length and there are hotkeys for the shell commands.

Super cool tip!

**Comment by 7flash** [comment]
> Hi! I would like to share here my current workflow I found myself comfortable with.
> 
> [7flash/helix-chatgpt](https://github.com/7flash/helix-chatgpt)
> 
> Notice, it works especially well with Warp Terminal, where I created a workflow/button to execute the script, but as well you can define bash function etc.
> 
> What happens when you run the script, it opens a new file with helix where you can write your prompt, and I found it being more comofrtable than any existing UI.
> 
> * You can separate user and assistant messages with |user| and |assistant| separators directly in your prompt
> * You can reference local files with file:/ references and web links with http[s]:// and their content will be embedded into the prompt
> * Now, you obviously won't have to embed the whole content of given webpage into your prompt and therefore, you can specify CSS selector after # hashtag in its path.
> * As well, you won't have to embed full source files, so you can reference specific sections using special comments starting with "."
> 
> Just sharing here what works for me, but there isn't good documentation yet, so please feel free to contribute.

Here is upgraded version: https://github.com/7flash/gurrai

**Comment by hchasens** [comment]
I'd suggest modifying [Continue](https://github.com/continuedev/continue/). It's under the Apache License and is one of the most mature open-source AI plugins. Since it's also on VSCode and JetBrains it should be well abstracted. It supports a number of backends and features.

Perhaps Helix can look into teaming with Continue sometime in the future?

**Comment by sbrow** [comment]
For anyone looking for a terminal friendly alternative to [cursor](https://www.cursor.com/), one of my co-workers just shared https://aider.chat/ with me. You might be able to make a helix plugin / language server for it? :man_shrugging:

**Reply by David-Else** [reply]
aider is considered the best, I can recommend it too. It does seem to be amazing for some things, and rubbish for others, so give it a go, it needs a bit of experimentation to find out how it works.

**Comment by dpc** [comment]
I really want to get some AI Coding assistant in Helix

I tried `lsp-ai` and while it seems to have a lot of stuff done under the hood it is actually poorly integrated with Helix itself: it requires writing own configuration to extract code from reponses, etc.

Then I tried `helix-gpt` and this one seems to have the UX integration with Helix largely figured out, [but the author/maintainer seems no longer active](https://github.com/leona/helix-gpt/discussions/74). The latest release seemed to have some severe issues with timeouts that remained unfixed for a while now. I was able to patch it up in a fork, and  it works quite OKish for me now. If any developers would be interested in stepping up, I think it could easily be a go-to solution for Helix.

**Reply by kirawi** [reply]
I'd actually be interested in working on a fork implementing MCP since I see it becoming the standardized interface for editors, but uh (looks at the major features I'm years behind on) I don't think it'll be happening :P

**Reply by devmanuelli** [reply]
@dpc see here https://github.com/helix-editor/helix/pull/14876

**Comment by chiefMarlin** [comment]
I added 2 new commands in pr #12590 that makes it easier to use shell commands like `aichat` for suggestion and completion.

TLDR;
This PR introduces two new commands:

`pipe-append`
Takes selected text and pipes it to a shell command.
Appends the shell command's output on a new line immediately after the selected text.

`pipe-inline-append`
Works similarly but inserts the shell command's output directly at the cursor position, rather than on a new line.

Intent of this is to add flexibility in using shell commands for text completion and insertion.
For example using shell commands to complete code/text.
`:pipe-append aichat Given the current context, suggest next line. Output only text, no explanations:`
`:pipe-append-inline aichat Given the current context, finish this line. Output only text, no explanations:`


Adding keybindings for completion
```
"⁄" = ":pipe-inline-append aichat \"Given the current context, finish this line. Output only text, no explanations:\""
"€" = ":pipe-append aichat \"Given the current context, suggest next line. Output only text, no explanations:\""
```

**Reply by ristomatti** [reply]
(I'll be the dork stating the obvious).

One would likely want to use aichat's role feature to pass in the prompt. E.g.

```toml
[keys.normal]
"tab" = ":pipe-append aichat --role helix-context"
```

and then in `$HOME/.config/aichat/roles/helix-context.yaml`:

```yaml
Given the current context, suggest next line. Output only text, no explanations:
```

This way you can write a longer prompt w/o cluttering the config (if needed).

**Reply by kihaya** [reply]
This method is very versatile and simple. Any ideas on how to send additional context to aichat, such as a list of the editor's open buffers? (Like Github Copilot plugin in VS Code)

**Reply by chiefMarlin** [reply]
> This method is very versatile and simple. Any ideas on how to send additional context to aichat, such as a list of the editor's open buffers? (Like Github Copilot plugin in VS Code)

I think once this PR #11164 gets merged then we would be able to fetch the file/project files and provide that as context in addition to the prompt and selection.
https://github.com/helix-editor/helix/pull/11164

**Comment by webdev23** [comment]
You have to add support for a text streaming pipeline.

**Comment by iocron** [comment]
Any new updates on this?

**Reply by devmanuelli** [reply]
@iocron checkout my pr https://github.com/helix-editor/helix/pull/14876
